File: herocats_extras/Accepted.java
import java.util.*;

public class Accepted {
    static int[][] missions;
    static int[][][] cache;

    public static int performMissions(int moneyCap, int timeCap, int missionI) {
        if (moneyCap == 0 || timeCap == 0 || missionI == 0) {
            return 0;
        }

        if (cache[missionI][timeCap][moneyCap] != -1) {
            return cache[missionI][timeCap][moneyCap];
        }

        if (missions[missionI-1][0] > moneyCap || missions[missionI-1][1] > timeCap) {
            int result = performMissions(moneyCap, timeCap, missionI - 1);
            cache[missionI][timeCap][moneyCap] = result;
            return result;
        }

        int cost = missions[missionI-1][0];
        int timeConsumed = missions[missionI-1][1];
        int rescues = missions[missionI-1][2];

        int performRescue = performMissions(moneyCap - cost, timeCap - timeConsumed, missionI - 1) + rescues;
        int dontRescue = performMissions(moneyCap, timeCap, missionI - 1);
        int result = Math.max(performRescue, dontRescue);

        cache[missionI][timeCap][moneyCap] = result;
        return result;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int M = sc.nextInt();
        int T = sc.nextInt();
        int N = sc.nextInt();

        missions = new int[N][3];
        cache = new int[N+1][T+1][M+1];
        for (int[][] layer2d : cache) {
            for (int[] layer1d : layer2d) {
                Arrays.fill(layer1d, -1);
            }
        }

        for (int i = 0; i < N; i++) {
            int cost = sc.nextInt();
            int timeConsumed = sc.nextInt();
            int rescues = sc.nextInt();
            missions[i][0] = cost;
            missions[i][1] = timeConsumed;
            missions[i][2] = rescues;
        }

        int rescues = performMissions(M, T, N);
        System.out.println(rescues);
    }
}

File: herocats/submissions/accepted/accepted.py
#! /usr/bin/env python3
import sys

sys.setrecursionlimit(10 ** 6)
CACHE = {}


def perform_missions(money_cap, time_cap, mission_i):
    # Base case 1
    if money_cap == 0 or time_cap == 0 or mission_i == 0:
        return 0

    if (mission_i, time_cap, money_cap) in CACHE and CACHE[(mission_i, time_cap, money_cap)] != -1:
        return CACHE[(mission_i, time_cap, money_cap)]

    # Base case 2
    if missions[mission_i - 1][0] > money_cap or missions[mission_i - 1][1] > time_cap:
        result = perform_missions(money_cap, time_cap, mission_i - 1)
        CACHE[(mission_i, time_cap, money_cap)] = result
        return result

    # Recursive loop
    cost, time_consumed, rescues = missions[mission_i - 1]

    perform_rescue = perform_missions(money_cap - cost, time_cap - time_consumed, mission_i - 1) + rescues
    dont_rescue = perform_missions(money_cap, time_cap, mission_i - 1)
    result = max(perform_rescue, dont_rescue)

    CACHE[(mission_i, time_cap, money_cap)] = result
    return result


if __name__ == '__main__':
    # start_time = time.time()
    M, T, N = map(int, input().split())
    missions = []
    for i in range(N):
        missions.append(tuple(map(int, input().split())))  # Map with move span
    rescues = perform_missions(M, T, len(missions))
    print(rescues)
    # end_time = time.time()
    # print("Total execution time: ", end_time - start_time, "s")
